mis_builder_cash_flow/report/mis_cash_flow.py

```diff
@@ -0,0 +1,129 @@
+from odoo import api, fields, models, tools
+from psycopg2.extensions import AsIs
+
+
+class MisCashFlow(models.Model):
+
+    _name = 'mis.cash_flow'
+    _description = 'MIS Cash Flow'
+    _auto = False
+
+    # line_type = fields.Char()
+    line_type = fields.Selection(
+        [('forecast_line', 'Forecast Line'), ('move_line', 'Journal Item')],
+        index=True,
+        readonly=True,
+    )
+    name = fields.Char(
+        readonly=True,
+    )
+    account_id = fields.Many2one(
+        comodel_name='account.account',
+        string='Account',
+        auto_join=True,
+        index=True,
+        readonly=True,
+    )
+    move_line_id = fields.Many2one(
+        comodel_name='account.move.line',
+        string='Journal Item',
+        auto_join=True,
+        readonly=True,
+    )
+    company_id = fields.Many2one(
+        comodel_name='res.company',
+        string='Company',
+        auto_join=True,
+        readonly=True,
+        index=True,
+    )
+    credit = fields.Float(
+        readonly=True,
+    )
+    debit = fields.Float(
+        readonly=True,
+    )
+    date = fields.Date(
+        readonly=True,
+        index=True,
+    )
+    reconciled = fields.Boolean(
+        readonly=True,
+    )
+    full_reconcile_id = fields.Many2one(
+        'account.full.reconcile',
+        string="Matching Number",
+        readonly=True,
+        index=True,
+    )
+    user_type_id = fields.Many2one(
+        'account.account.type',
+        auto_join=True,
+        readonly=True,
+        index=True,
+    )
+
+    @api.model_cr
+    def init(self):
+        query = """
+            SELECT
+                -- we use neagative id to avoid duplicates and we don't use
+                -- ROW_NUMBER() because the performance was very poor
+                -aml.id as id,
+                CAST('move_line' AS varchar) as line_type,
+                aml.id as move_line_id,
+                aml.account_id as account_id,
+                aml.debit as debit,
+                aml.credit as credit,
+                aml.reconciled as reconciled,
+                aml.full_reconcile_id as full_reconcile_id,
+                aml.company_id as company_id,
+                aml.user_type_id as user_type_id,
+                aml.name as name,
+                CASE
```

Comment:
I am testing on a DB with millions of move lines, and it is unusable, I believe due to this case.

Looking at the query I see that `date_maturity` is already a required field. So checking it if it is not null does not seem to make sense. https://github.com/odoo/odoo/blob/11.0/addons/account/models/account_move.py#L450. So we can assume that for receivables and payables date_maturity is what needs to be fetched.

So, Instead of the case, I would think that two queries would work better, as:
```
SELECT
                -- we use neagative id to avoid duplicates and we don't use
                -- ROW_NUMBER() because the performance was very poor
                -aml.id as id,
                CAST('move_line' AS varchar) as line_type,
                aml.id as move_line_id,
                aml.account_id as account_id,
                aml.debit as debit,
                aml.credit as credit,
                aml.reconciled as reconciled,
                aml.full_reconcile_id as full_reconcile_id,
                aml.company_id as company_id,
                aml.user_type_id as user_type_id,
                aml.name as name,
                **aml.date_maturity as date,**                
            FROM account_move_line as aml            
            JOIN account_account_type aat ON aml.user_type_id = aat.id
            WHERE aat.type IN ('receivable', 'payable')
            UNION ALL
            SELECT
                -- we use neagative id to avoid duplicates and we don't use
                -- ROW_NUMBER() because the performance was very poor
                -aml.id as id,
                CAST('move_line' AS varchar) as line_type,
                aml.id as move_line_id,
                aml.account_id as account_id,
                aml.debit as debit,
                aml.credit as credit,
                aml.reconciled as reconciled,
                aml.full_reconcile_id as full_reconcile_id,
                aml.company_id as company_id,
                aml.user_type_id as user_type_id,
                aml.name as name,
               am.date as date
            FROM account_move_line as aml            
            JOIN account_account_type aat ON aml.user_type_id = aat.id
            WHERE aat.type NOT IN ('receivable', 'payable')                    
            UNION ALL
            SELECT
                fl.id as id,
                CAST('forecast_line' AS varchar) as line_type,
                Null as move_line_id,
                fl.account_id as account_id,
                CASE
                    WHEN fl.balance > 0
                    THEN fl.balance
                    ELSE 0.0
                END AS debit,
                CASE
                    WHEN fl.balance < 0
                    THEN -fl.balance
                    ELSE 0.0
                END AS credit,
                Null as reconciled,
                Null as full_reconcile_id,
                fl.company_id as company_id,
                aa.user_type_id as user_type_id,
                fl.name as name,
                fl.date as date
            FROM mis_cash_flow_forecast_line as fl
            JOIN account_account aa ON aa.id = fl.account_id
```

Reply:
Great Jordi,
I've directly change to use maturity date for all the move lines because actually it can be useful on liquidity lines where you may have a future date (for eg a check with a date_maturity in the future), on most cases date_maturity = date.

Can you test performance with this change?
If performance is still poor try removing everythng on the query after "UNION ALL", of course you are not going to have the forecast lines, just to get if the issue is there...